### PR TITLE
[HttpClient] Add method to set response factory in mock client

### LINKED ISSUE #40620: [HttpClient] Allow MockHttpClient $responseFactory to be modifiable though accessors
**Description**
What I'm trying to achieve is the following : dynamically setting up a MockHttpClient replacement for the HttpClientInterface typehint inside my service when running Behat tests.  
The idea is to be able to run a scenarii that will span the entire call stack (from the controller, all the way down to the HttpClient call and back) by setting the HttpClientInterface alias to a MockHttpClient service.  
Each test should be able to define its own set of responses to be returned by the MockHttpClient upon call.

At the moment, this does not seem to be possible as the responseFactory needs to be injected into the MockHttpClient constructor and there are no accessors to change it afterwards.

It would be nice to be able to call a setResponseFactory from within a Behat test (or any other test type) to set the specific list of responses expected for the current test like this : 
```php
$mockHttpClient->setResponseFactory({My array of ResponseInterfaces});
```

[Note] 
I try using in test environment the following without success
```yaml
services:
    _defaults:
        autowire: true
        autoconfigure: true
        public: true
```

And in the behat test context
```php
$this->kernel
            ->getContainer()
            ->set(HttpClientInterface::class, new MockHttpClient({My array of ResponseInterfaces}));
```

If there is a way to achieve this, without having to create specific classes for test purposes only, please share.

### PR DESCRIPTION
**Pull request by greeflas**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no 
| Tickets       | Fix #40620 
| License       | MIT
| Doc PR        | symfony/symfony-docs#15883

### CODE REVIEW

File: src/Symfony/Component/HttpClient/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+5.4
+---
+
+* added `MockHttpClient::setResponseFactory()` method to be able to set response factory after client creating
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/CHANGELOG.md]: missing space at start of line
"added" should be replaced by "Add"

File: src/Symfony/Component/HttpClient/MockHttpClient.php [hunk 2]
```diff
@@ -47,7 +56,6 @@ public function __construct($responseFactory = null, ?string $baseUri = 'https:/
         }
 
         $this->responseFactory = $responseFactory;
-        $this->defaultOptions['base_uri'] = $baseUri;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/MockHttpClient.php:50]: looks like a mistake, isn't it?

**Review by nicolas-grekas — APPROVED** [review]
(with minor comments)

File: src/Symfony/Component/HttpClient/CHANGELOG.md (hunk 1 above)
Comment (greeflas) [reply src/Symfony/Component/HttpClient/CHANGELOG.md]: Fixed

File: src/Symfony/Component/HttpClient/MockHttpClient.php [hunk 3]
```diff
@@ -35,6 +35,15 @@ class MockHttpClient implements HttpClientInterface
...
     public function __construct($responseFactory = null, ?string $baseUri = 'https://example.com')
+    {
+        $this->setResponseFactory($responseFactory);
+        $this->defaultOptions['base_uri'] = $baseUri;
+    }
+
+    /**
+     * @param callable|callable[]|ResponseInterface|ResponseInterface[]|iterable|null $responseFactory
+     */
+    public function setResponseFactory($responseFactory)
```
Comment (lmillucci) [inline src/Symfony/Component/HttpClient/MockHttpClient.php]: I think you should add `: void` as return type

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/MockHttpClient.php]: oh, or better, make this a wither: `withResponseFactory(): static`?

Comment (norkunas) [reply src/Symfony/Component/HttpClient/MockHttpClient.php]: How `wither` would help in unit tests when you instantiated a mock client in a `setUp` method and you need to change the factory in a specific test?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/MockHttpClient.php]: good point :)

Comment (greeflas) [reply src/Symfony/Component/HttpClient/MockHttpClient.php]: Added return type.

**Comment by derrabus** [comment]
Thank you for this PR. I agree with the changes, but I think we should add a test.

**Review by derrabus — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @greeflas.

**Pull request merged**